ngUpgrade AoT

petebacondarwin opened the pull request:
**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

petebacondarwin commented on karma-js.conf.js:
```diff
@@ -23,7 +23,7 @@ module.exports = function(config) {
       'node_modules/core-js/client/core.js',
       // include Angular v1 for upgrade module testing
-      'node_modules/angular/angular.min.js',
+      'node_modules/angular/angular.js',
```
TODO: remember to revert this change

alexeagle replied:
can revert it?

petebacondarwin replied:
@mhevery said it can stay un-minified

mhevery commented on modules/@angular/upgrade/index.ts:
```diff
@@ -12,5 +12,6 @@
  * Entry point for all public APIs of the upgrade package.
  */
 export * from './src/upgrade';
+export * from './src/upgrade_module';
```
Don't export all. Instead list which items specifically you want to export.  Otherwise `$injectorFactory` and friends will be part of public API.

petebacondarwin replied:
Right, OK. I'll fix up.

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
```diff
@@ -0,0 +1,147 @@
+@NgModule({
+  providers: [
+    { provide: '$injector', useFactory: $injectorFactory },
+    { provide: '$rootScope', useFactory: $rootScopeFactory, deps: ['$injector']},
```
@chuckjaz Notice how we have to explicitly list `$rootScopeFactory` and `$compileFactory` in future we will have many more. It would be nice to just have one function which does it all, but we get an Error:

```
Metadata collected contains an error that will be reported at runtime:
   Function calls are not supported.
   Consider replacing the function or lambda with a reference to an exported function
```

I thought we supported this, but apparently not. Could you work with @petebacondarwin and help him resolve it?

chuckjaz replied:
@mhevery Sure.

@petebacondarwin Can I see the function that you are trying to use? It might be easy to transform.

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
```diff
@@ -0,0 +1,147 @@
+  providers: [
+    { provide: '$injector', useFactory: $injectorFactory },
+    { provide: '$rootScope', useFactory: $rootScopeFactory, deps: ['$injector']},
+    { provide: '$compile', useFactory: $compileFactory, deps: ['$injector']}
```
Should we re-export all of the `$*` items?

petebacondarwin replied:
There are lots of them...
Perhaps we could just provide a recipe in the docs for people to add their own to their subclass of UpgradeModule?

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
```diff
@@ -0,0 +1,147 @@
+/**
+ * UpgradeModule is the base class for the module that will contain all the
+ * information about what Angular providers and component are to be bridged
+ * between Angular 1 and Angular 2+
```
We will have to update this doc with an example. We now require that full API documentation will be included at the time of submission of the PR.

//cc @IgorMinar

petebacondarwin replied:
These docs were mostly placeholders to help with my understanding.
We can add more as the work stabilizes

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
````diff
@@ -0,0 +1,147 @@
+   * Usage:
+   *
+   * ```
+   * angular1Module.factory('someService', ng2ProviderFactory(SomeService))
````
turn these into running examples. See: https://github.com/angular/angular/pull/11773 for how the docs should look / work.

petebacondarwin replied:
will do as things stabilize

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
````diff
@@ -0,0 +1,147 @@
+   * ```
+   */
+  static ng2ProviderFactory(token: any) {
+    return ['ng2Injector', (i: Injector) => i.get(token)];
````
Maybe extract the strings into constants?

petebacondarwin replied:
there are actually constants that I should be using in `./constants.ts`.

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
````diff
@@ -0,0 +1,147 @@
+   * })
+   * ```
+   */
+  static ng1ServiceProvider({provide, ng1Token}: {provide: any, ng1Token: any}): FactoryProvider {
````
We should think about the naming.  `ng2ProviderFactory` vs `ng1ServiceProvider` seem to different. Think about `downgrade`/`upgrade` word `ng1`/`ng1` name. 

Ideas:
-  `downgradeNg2Injectable` vs `upgradeNg1Injectable`
- `wrapAsNg1Factory` vs `wrapAsNg2Provider`
- `createNg1FactoryFromNg2` vs `createNg2ProviderFromNg1`

robwormald replied:
ideally not Ng2 anything, as it'll be Ng3 anything in a few months...

petebacondarwin replied:
That is the tricky bit eh @robwormald  :-P

* `createDowngradeProvider` vs `createUpgradeFactory`

?

petebacondarwin replied:
Perhaps we drop ng2 throughout and have:

* `wrapAsProvider` vs `wrapAsNg1Factory`

And then have things like

* `ref.instance.injector` : Angular 2 injector
vs
* `ref.instance.ng1Injector` : Angular 1 injector

mhevery replied:
+1 for not having `ng2` anything.

I vote for - `ng1` prefix for `ng1` but no prefix for `ng2` (just as @petebacondarwin) suggested.

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
```diff
@@ -0,0 +1,147 @@
+    (<any>directiveFactory).$inject = ['$injector', '$parse', directiveFactory];
+    return directiveFactory;
+
+    function directiveFactory(
```
this seems like cut and paste from ngUpgrade. Could we somehow refactor and keep the codebase shared? or do you think it is not worth it? (just asking)

petebacondarwin replied:
Funnily enough I was just doing some tidying up on this.
I would like to get tests passing before doing refactoring with regards to the rest of the ngUpgrade codebase.

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
```diff
@@ -0,0 +1,147 @@
+    function directiveFactory(
+        ng1Injector: angular.IInjectorService,
+        parse: angular.IParseService): angular.IDirective {
+      const NG2_INJECTOR = 'ng2.Injector';
```
make sure we have consistent naming. either `ng2.foo` or `ng2Foo`

mhevery replied:
I see the point with `ng2`. So maybe we should just do `$injector` and `ng.injector`? as the two ways. Renaming it would be a breaking change, even thought it is not part of the public API so it is unlikely that anyone would be using it. But sounds like the ngUpgradeAoT should use `ng.Injector`

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
```diff
@@ -0,0 +1,147 @@
+    this.ngZone = ng2Injector.get(NgZone);
+  }
+
+  ngDoBootstrap() {}
```
add comment as to which this method is here.

petebacondarwin replied:
strangely I did have a comment here but it seems to have failed to make it into the commit that was pushed.

mhevery commented on modules/@angular/upgrade/src/upgrade_module.ts:
```diff
@@ -0,0 +1,147 @@
+
+    // Only add the config if it is there
+    // QUESTION? Do we really need this param?
+    if (config) {
```
why is this necessary? Wouldn't `angular.bootstrap` take care of it?

mhevery commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -0,0 +1,1103 @@
+
+    it('should have angular 1 loaded', () => expect(angular.version.major).toBe(1));
+
+    it('should instantiate ng2 in ng1 template and project content', async(() => {
```
@juliemr we are having a hard time making the `async` work here. We are getting timeout. Could you work with @petebacondarwin on figuring out why it is failing to stabilize?

petebacondarwin replied:
I think this may have resolved itself...

mhevery commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -0,0 +1,1103 @@
+          html('<div>{{ \'ng1[\' }}<ng2>~{{ \'ng-content\' }}~</ng2>{{ \']\' }}</div>');
+
+      platformBrowserDynamic().bootstrapModule(Ng2Module).then((ref) => {
+        ref.instance.bootstrapNg1(element, [ng1Module.name]);
```
Why do it here, rather than in the `Ng2Module` constructor?

petebacondarwin replied:
Moving the call to `bootstrapNg1` outside the constructor makes the custom UpgradeModule more reusable. It doesn't tie it into always bootstrapping on a specific element.

mhevery replied:
I see, good point.

mhevery reviewed (changes requested):
Awesome start and great progress!

petebacondarwin commented on modules/@angular/upgrade/src/aot/ng1_service_provider.ts:
```diff
@@ -0,0 +1,19 @@
+ */
+export function ng1ServiceProvider({provide, ng1Token}: {provide: any, ng1Token: any}): FactoryProvider {
+  return {provide: provide, useFactory: (i: angular.IInjectorService) => i.get(ng1Token), deps: ['$injector']};
+}
```
shouldn't this helper fall foul of the same problems with not being allowed to use a non-exported named function as a factory?
cc @mhevery & @chuckjaz

petebacondarwin commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -255,38 +257,47 @@ export function main() {
-      //          }
-      //        };
-      //      });
+      it('should properly run cleanup when ng1 directive is destroyed', async(() => {
```
@gkalpak  can you take a look and see if this test makes sense. The previous commented out test that this replaces is very different and in I think doesn't even work properly.

gkalpak replied:
The test makes sense. I think the previosu test tried to do the same thing.
One thing I noticed is that the previosu test calls `ref.dispose()` at the end. Is it necessary? What was the purpose?

petebacondarwin commented:
@gkalpak please do the honors for the googlebot

gkalpak commented on modules/@angular/upgrade/src/angular_js.ts:
```diff
@@ -51,7 +51,7 @@ export interface IRootScopeService {
 }
 export interface IScope extends IRootScopeService {};
-export interface IAngularBootstrapConfig {}
+export type IAngularBootstrapConfig = IInjectable;
```
But it is not the same. `IAngularBootstrapConfig` is the configuration that can be passed to `angular.bootstrap` and can contain `stringDi: true/false` for example. (It can also contain `debugInfoEnabled`, but that is undocumented and used internally.) Essentially, it is just an object with specific properties.

gkalpak replied:
So, it should probably be changed to:

```ts
export interface IAngularBootstrapConfig {
  strictDi?: boolean;
}
```

petebacondarwin commented:
Oops

On 13 Oct 2016 22:04, "Georgios Kalpakas" <notifications@github.com> wrote:

> *@gkalpak* commented on this pull request.
> ------------------------------
>
> In modules/@angular/upgrade/src/angular_js.ts
> <https://github.com/angular/angular/pull/12239>:
>
> > @@ -51,7 +51,7 @@ export interface IRootScopeService {
>    [key: string]: any;
>  }
>  export interface IScope extends IRootScopeService {};
> -export interface IAngularBootstrapConfig {}
> +export type IAngularBootstrapConfig = IInjectable;
>
> So, it should probably be changed to:
>
> export interface IAngularBootstrapConfig {
>   strictDi?: boolean;
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/12239>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA9J-80qsnq9YruZZabnZy49Ezy2RyVks5qzpzTgaJpZM4KUxSw>
> .
>

gkalpak commented:
@googlebot, I confirm that I'm okay with my commits being contributed to Google :smiley:

gkalpak commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -342,6 +345,9 @@ export function main() {
           adapter.bootstrapNg1(element, [ng1Module.name]);
+          // the fact that the body contains the correct text means that the
+          // downgraded component was able to access the moduleInjector
+          // (since there is no other injector in this system)
```
Still not convinced 😛

petebacondarwin replied:
Trouble is until we have `upgradeComponent` working we can't do better than this test

mhevery commented on modules/@angular/upgrade/src/aot/ng1_module.ts:
```diff
@@ -0,0 +1,41 @@
+ */
+@NgModule({
+  providers: [
+    Ng1Adapter,
```
Why is this broken up to a separate class? Why no just part of the module? I am on the fence, and wanted to understand your reasoning.

mhevery commented on modules/@angular/upgrade/src/aot/ng1_module.ts:
```diff
@@ -0,0 +1,41 @@
+export class Ng1Module {}
+
+let tempGlobalNg1Injector: angular.IInjectorService = null;
+/** @internal */
```
I don't think `@internal` provides any value. Remove? (same as below)

mhevery commented on modules/@angular/upgrade/src/constants.ts:
```diff
@@ -7,10 +7,11 @@
 
 export const NG2_COMPILER = 'ng2.Compiler';
-export const NG2_INJECTOR = 'ng2.Injector';
+export const UPGRADE_MODULE_INJECTOR = 'ng2.Injector';
```
`export const COMPONENT_INJECTOR = 'componentInjector';`

petebacondarwin replied:
So similar question to earlier. This constant could refer to the injector on a component or to the "root" injector on the module. Should we have two constants?

mhevery replied:
Good question. My initial reaction is that those two things are separate, but then again it is not like you can inject module injector in Ng2. 

But the difference is that `element.data()` will never return a module injector only component injector. So I think having two different names may be easier for the reader of the code.

Fine either way.

petebacondarwin replied:
I think we are not using these constants in the new AoT code.

mhevery commented on modules/@angular/upgrade/src/constants.ts:
```diff
@@ -7,10 +7,11 @@
 export const NG2_COMPILER = 'ng2.Compiler';
-export const NG2_INJECTOR = 'ng2.Injector';
+export const UPGRADE_MODULE_INJECTOR = 'ng2.Injector';
 export const NG2_COMPONENT_FACTORY_REF_MAP = 'ng2.ComponentFactoryRefMap';
```
not neede

petebacondarwin replied:
Except that it is still used by the JIT `UpgradeAdapter`

mhevery commented on modules/@angular/upgrade/src/constants.ts:
```diff
@@ -7,10 +7,11 @@
-export const NG2_INJECTOR = 'ng2.Injector';
+export const UPGRADE_MODULE_INJECTOR = 'ng2.Injector';
 export const NG2_COMPONENT_FACTORY_REF_MAP = 'ng2.ComponentFactoryRefMap';
 export const NG2_ZONE = 'ng2.NgZone';
```
not needed

mhevery commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -0,0 +1,1197 @@
+          entryComponents: [Ng2Component]
+        })
+        class Ng2Module  {
+          ngDoBootstrap() {}
```
nice thing about extending was that one did not have to write `ngDoBootstrap` method.

mhevery commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -0,0 +1,1197 @@
+        // the ng1 app module that will consume the downgraded component
+        const ng1Module = angular.module('ng1', [])
+          // create an ng1 facade of the ng2 component
+          .directive('ng2', downgradeComponent({ selector: 'ng2', type: Ng2Component }));
```
Why do we have to pass in the `selector`? it should not be needed.

Also can we rename `type` to `component`?

petebacondarwin replied:
You are right. I was being lazy because I was reusing `DowngradeNg2ComponentAdapter` from the JIT version which accepts a `ComponentInfo` type, which requires the selector. 

In fact the only use of `info.selector` in that class is when reporting an error.

We should refactor this and remove the selector.

mhevery commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -0,0 +1,1197 @@
+            html('<div>{{ \'ng1[\' }}<ng2>~{{ \'ng-content\' }}~</ng2>{{ \']\' }}</div>');
+
+        platformBrowserDynamic().bootstrapModule(Ng2Module).then((ref) => {
+          const adapter = ref.injector.get(Ng1Adapter) as Ng1Adapter;
```
By having `Ng1Adapter` as a separate class, there is more things to import. Still on the fence about that.

petebacondarwin replied:
Let's mull it over during the weekend. Happy to revert if it still feels wrong.

mhevery reviewed (changes requested):
Excellent progress sir!

petebacondarwin commented on modules/@angular/upgrade/src/angular_js.ts:
```diff
@@ -6,20 +6,30 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+export type Ng1Token = string;
```
I was looking forward to if/when we might allow non-string tokens in ng1

petebacondarwin commented on modules/@angular/upgrade/src/angular_js.ts:
```diff
@@ -6,20 +6,30 @@
 
+export type Ng1Token = string;
+
+export interface IAnnotatedFunction extends Function {
```
Well I didn't know about https://www.npmjs.com/package/@types/angular.
@gkalpak and I probably wasted a bit of time over the last few days tweaking this file.

petebacondarwin replied:
I think we would need to do some work with that repository as I am not entirely happy with all of the hand crafted types there.

petebacondarwin commented on modules/@angular/upgrade/src/aot/downgrade_component.ts:
```diff
@@ -0,0 +1,51 @@
+
+    return {
+      restrict: 'E',
+      require: '?^' + UPGRADE_MODULE_INJECTOR,
```
OK, do we still leave the UPGRADE_MODULE_INJECTOR constant for when we are accessing the top level injector? E.g. `$injector.get(UPGRADE_MODULE_INJECTOR)`

mhevery replied:
correct

mhevery commented on modules/@angular/upgrade/test/upgrade_aot_spec.ts:
```diff
@@ -0,0 +1,1197 @@
+
+        @NgModule({
+          declarations: [Ng1Component, Ng2Component],
+          imports: [BrowserModule]
```
`Ng1Module` => `UpgradeModule`

alexeagle commented on modules/@angular/upgrade/src/metadata.ts:
```diff
@@ -27,8 +27,8 @@ export interface AttrProp {
 export interface ComponentInfo {
```
not part of your change - but the import of DirectiveResolver above causes @angular/upgrade to depend on @angular/compiler so my AoT example is still getting the compiler at runtime, which makes the app big. Maybe a Closure Compiler issue but it's easier to trim the dependency.
@mhevery this suggests the aot version of upgrade should be a separate package so we can avoid the reflection dependencies?

petebacondarwin committed:
chore(tslint.json): semicolon rule expects an array

petebacondarwin committed:
feat(ngUpgrade): add support for AoT compiled upgrade applications

This commit introduces a new API to the ngUpgrade module, which is compatible
with AoT compilation. Primarily, it removes the dependency on reflection
over the Angular 2 metadata by introducing an API where this information
is explicitly defined, in the source code, in a way that is not lost through
AoT compilation.

This commit is a collaboration between @mhevery (who provided the original
design of the API); @gkalpak & @petebacondarwin (who implemented the
API and migrated the specs from the original ngUpgrade tests) and @alexeagle
(who provided input and review).

This commit is an starting point, there is still work to be done:

* add more documentation
* validate the API via internal projects
* align the ngUpgrade compilation of A1 directives closer to the real A1
  compiler
* add more unit tests
* consider support for async `templateUrl` A1 upgraded components

Closes #12239

alexeagle commented on tools/public_api_guard/upgrade/index.d.ts:
```diff
@@ -19,3 +25,20 @@ export declare class UpgradeAdapterRef {
+
+/** @experimental */
+export declare class UpgradeModule {
+    $injector: angular.IInjectorService;
```
this only works if you have this local angular1 typings to make it match.
Real users with angular 1 and angular 2 typings will find a structural mismatch here. Needs to be `any`

petebacondarwin replied:
We agreed to wait for it to break stuff

petebacondarwin commented on modules/@angular/upgrade/test/aot/test_helpers.ts:
```diff
@@ -0,0 +1,39 @@
+    upgrade.bootstrap(element, [ng1Module.name]);
+    return upgrade;
+  });
+}
```
@mhevery - we should consider adding this helper to the public API
